walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,147 @@
+import java.util.List;
+import java.util.Set;
 
+/**
+ * Class that handles finding eligible time slot.
+ */
 public final class FindMeetingQuery {
+  /**
+   * Gets the available time slot for new meeting given existing meeting events.
+   * @param events Existing meeting events, assuming all events end before EOD.
+   * @param request Request to set up a new meeting with duration not exceeding the whole day.
+   * @return If one or more time slots exists so that both mandatory and optional attendees can attend,
+   *         returns those time slots. Otherwise, returns the time slots that fit just the mandatory attendees.
+   *         Returns empty collection if invalid input.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long durationInMinutes = request.getDuration();
+    if (!isValidParams(events, durationInMinutes)) {
+      return ImmutableList.of();
+    }
+
+    Set<String> attendees = new HashSet<>(request.getAttendees());
+    Set<String> attendeesWithOptional = new HashSet<>();
+    attendeesWithOptional.addAll(request.getAttendees());
+    attendeesWithOptional.addAll(request.getOptionalAttendees());
+
+    // No attendees at all.
+    if (attendeesWithOptional.isEmpty()) {
+      return ImmutableList.of(TimeRange.WHOLE_DAY);
+    }
+
+    ImmutableList<TimeRange> occupiedTimeRange = getOccupiedTimeRange(attendees, events);
+    ImmutableList<TimeRange> occupiedTimeRangeWithOptional = getOccupiedTimeRange(attendeesWithOptional, events);
+
+    ImmutableList<TimeRange> availableMeetings = getAvailableTimeRange(occupiedTimeRangeWithOptional, durationInMinutes);
+
+    // Has time slots that satisfy both mandatory and optional attendees.
+    if (!availableMeetings.isEmpty()) {
+      return availableMeetings;
+    }
+
+    // No mandatory attendees.
+    if (attendees.isEmpty()) {
+      return ImmutableList.of();
+    }
+
+    return getAvailableTimeRange(occupiedTimeRange, durationInMinutes);
+  }
+
+  private boolean isValidParams(Collection<Event> events, long duration) {
+    // TODO: check for events that all events not crossing EOD.
+
+    if (duration > TimeRange.WHOLE_DAY.duration() || duration <= 0) {
+      return false;
+    }
+
+    return true;
+  }
+
+  private ImmutableList<TimeRange> getOccupiedTimeRange(Set<String> attendees, Collection<Event> events) {
+    ImmutableList.Builder<TimeRange> occupiedTimeRange = ImmutableList.builder();
+    for (Event event: events) {
+      if (hasIntersection(event.getAttendees(), attendees)) {
+        occupiedTimeRange.add(event.getWhen());
+      }
+    }
+    return occupiedTimeRange.build();
+  }
+
+  private ImmutableList<TimeRange> getAvailableTimeRange(List<TimeRange> occupiedTimeRange, long duration) {
+    // Assumes 0 < duration <= WHOLE_DAY
+
+    ImmutableList.Builder<TimeRange> availableMeetings = ImmutableList.builder();
```

Comment:
nit: move this right before your for loop.

Reply:
Fixed